### ISSUE TITLE
Use plugin-based design for field definitions

### ISSUE THREAD

**Issue by Ichimonji10** [SatelliteQE/nailgun#214]
NailGun contains numerous bits of code like this:

``` python
        if _get_version(server_config) >= Version('6.1.1'):  # default: True
            self._fields.update({
                'default_content_view': entity_fields.OneToOneField(
                    ContentView
                ),
                'library': entity_fields.OneToOneField(LifecycleEnvironment),
            })
```

These bits of code are necessary because NailGun is (somewhat) intolerant of fields being declared incorrectly.

<ins>[edit: The following two examples are neat, but you can skip down to the "this is tiresome" section without missing too much.]</ins>

There's at least two especially obvious ways in which this is true. One way to trigger an exception is to pass in something silly when instantiating an entity:

``` python
>>> from nailgun import entities
>>> entities.Organization(not_a_real_field='foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "nailgun/entities.py", line 2586, in __init__
    super(Organization, self).__init__(server_config, **kwargs)
  File "nailgun/entity_mixins.py", line 387, in __init__
    .format(self._fields.keys(), kwargs.keys())
nailgun.entity_mixins.NoSuchFieldError: Valid fields are ['domain', 'description', 'default_content_view', 'library', 'compute_resource', 'id', 'subnet', 'realm', 'name', 'title', 'smart_proxy', 'hostgroup', 'label', 'environment', 'config_template', 'media', 'user'], but received ['not_a_real_field'] instead.
```

This is done to prevent mis-use of an entity by and end user, and it's sane behaviour. Another way to trigger an exception is for Satellite to not hand back enough data. That's how we end up with work-arounds like this:

``` python
    def create(self, create_missing=None):
        """Do extra work to fetch a complete set of attributes for this entity.

        For more information, see `Bugzilla #1230873
        <https://bugzilla.redhat.com/show_bug.cgi?id=1230873>`_.

        """
        return Organization(
            self._server_config,
            id=self.create_json(create_missing)['id'],
        ).read()
```

This is tiresome. As time goes on, we're going to add more and more conditional field definitions into every single entity. This will bloat the entity definitions and make them harder to maintain as time goes on. We should find a better solution.

One possibility is to use a plug-in based architecture. We currently have this layout:

```
$ tree nailgun
nailgun
├── client.py
├── config.py
├── entities.py
├── entity_fields.py
├── entity_mixins.py
└── __init__.py

0 directories, 6 files
```

Imagine that we added schema definitions, like so:

```
$ tree nailgun/
nailgun/
├── client.py
├── config.py
├── entities.py
├── entity_fields.py
├── entity_mixins.py
├── __init__.py
└── schemas
    ├── foreman
    │   └── 1.7.2.py
    └── satellite
        ├── 6.0.8.py
        ├── 6.1.0.py
        └── 6.1.1.py

3 directories, 10 files
```

The schemas could also be JSON or some other format, but the most straightforward approach would probably be for each schema to be a Python file containing field definitons just like what we have now. When an entity is instantiated, it can pick out the right schema and populate its `_meta` and `_fields` attributes.

Doing this would let us drop a huge amount of code (both magic constants and conditionals) from `nailgun/entities.py`, and we would be able to properly support Satellite, Foreman, RHCI and so on.

**Comment by Ichimonji10** [comment]
One possibility is to even put various bug work-arounds into the schema files, too. That would let us clear up `nailgun/entities.py` even more. I'm not sure how to handle that, though.

**Comment by omaciel** [comment]
@Ichimonji10 thank you for putting this together! @sthirugn had brought this up last week with me and we had a very interesting chat about it, but we did not at the time have any suggestions other than branching. The 'logic' behind using branch as an alternative to keep adding more conditional statements was that, once a release goes out the door we consider that the API is 'frozen' and won't really change. We would then have to create a build specific for that release (for those who are not using robottelo and want to play with a Satellite/Foreman instance on their own).

I particularly like the idea of being able to use **latest** NailGun without having to worry about what version of a Satellite/Foreman I have. The flexibility of writing one statement and have it just work against a 6.0.8, 6.1.0 and eventually 6.x.y system speaks for itself. But we do not, as you and @sthirugn  pointed out, want to get in the business of adding complex conditional statements to the code. At a first glance, your proposal of using versioned 'plugins' may aid us in keeping NailGun a very flexible tool.

So a question that comes to mind for this design is:

1. How exactly would "schema" files differ between versions? Would schema files be "cumulative" in the sense of inheritance? In other words, would 6.1.0.py inherit the schema from the prior version?

I'll spend a bit more time thinking about it and will add any comment/questions I may have  later.

**Comment by Ichimonji10** [comment]
> I particularly like the idea of being able to use **latest** NailGun without having to worry about what version of a Satellite/Foreman I have.

Spot on. This would be a huge benefit.

> How exactly would "schema" files differ between versions? Would schema files be "cumulative" in the sense of inheritance? In other words, would 6.1.0.py inherit the schema from the prior version?

I think it would be best if each schema was totally stand-alone. I don't want my changes to the schema for 6.0.7 to affect the 6.0.8, 6.1.1, 6.1.2, 6.2.0 etc schemas. Inheritance is fragile.

The down-side to having independent schemas is that there'll be a lot of duplication. But I think this can be mitigated with scripts. (Possibly several scripts, and possibly make targets too? We can start with just one or two simple scripts.) For example, we could have a script that scrapes an API, parses the JSON and spits out a basic schema file. That schema file would then need to be fleshed out by hand (the field types in particular would be a pain), but it's a start.

What would the schema look like? I'm thinking of something like this:

```python
schema = {
    'ActivationKey': {
        'fields': {
            'auto_attach': BooleanField(),
            'content_view': OneToOneField(ContentView),
            'description': StringField(),
            'environment': OneToOneField(LifecycleEnvironment),
            'host_collection': OneToManyField(HostCollection),
            'max_content_hosts': IntegerField(),
            'name': StringField(required=True),
            'organization': OneToOneField(Organization, required=True),
            'unlimited_content_hosts': BooleanField(),
        }
    },
    'Architecture': {…},
    …
```

If there's better approaches, of course, I'm very happy to hear them. One major problem with this approach is that we're working with raw Python code, and how in the world do you generate that? It may be better to make JSON schemas, or something else entirely.

**Comment by omaciel** [comment]
@Ichimonji10 thanks for the reply. Right now I don't have any suggestions as to a different approach to what a schema file would look like, but I do like what you proposed. I wish there was a way for the API generated by Satellite/Foreman to include a hint about the data type and/or relationship between different entities (ie. 1-to-1, 1-n, n-n), as it would improve our ability to generate these schemas or even allow us to generate them 'on the fly'!!! Do you remember if that python-foreman (the actual name escapes me now) does anything similar? I seem to recall that they were translating the JSON returned by the API into Pythonic objects...

**Comment by Ichimonji10** [comment]
[`python-foreman/foreman/definitions/`](https://github.com/david-caro/python-foreman/tree/master/foreman/definitions)

**Comment by Ichimonji10** [comment]
> I wish there was a way for the API generated by Satellite/Foreman to include a hint about the data type and/or relationship between different entities (ie. 1-to-1, 1-n, n-n), as it would improve our ability to generate these schemas or even allow us to generate them 'on the fly'!!!

That would be *fantastic*.

**Comment by omaciel** [comment]
> python-foreman/foreman/definitions/

Would you look at that :)

**Comment by Ichimonji10** [comment]
Here's another possible structure for a schema. It's JSON this time.

```json
"ActivationKey": {
    "fields": {
        "auto_attach": {"type": "BooleanField"},
        "content_view": {"type": "OneToOneField", "entity": "ContentView"},
        "description": {"type": "StringField"},
        "environment": {
            "entity": "LifecycleEnvironment",
            "type": "OneToOneField",
        },
        "host_collection": {
            "entity": "HostCollection",
            "type": "OneToManyField",
        },
        "max_content_hosts": {"type": "IntegerField"},
        "name": {"type": "StringField", "required": true},
        "organization": {
            "entity": "Organization",
            "required": true,
            "type": "OneToOneField",
        },
        "unlimited_content_hosts": {"type": "BooleanField"},
    }
},
"Architecture": {},
```

The up-shot of making a schema pure JSON is that you could create a `make validate-schemas` command. This would be *very* useful, given that people will be manually messing around with the schemas. The validator could not only validate that the schemas are valid JSON, but it could also validate that the schemas have only valid keywords. We could add other make targets too, like `make format-schemas`. This would let users mess around with the JSON any way they want, and the formatter would automatically re-write the JSON in a standard way.

**Comment by elyezer** [comment]
Other than the field definition would be required to have other information like workaround. For example many entities need to be wrapped in another hash, if that change from version to version?

If we use cumulative approach like migrations, then everytime nailgun start that should be computed. Using the generated api json reference would help (like in python-foreman) and we can generate entities dynamically, but we will not be able to handle the workarounds without additional information.

NailGun is smart enough to provide smart tools to work with the API by normalizing all differences and gotchas into a stable and clear API. We need to be careful here to not lose that as this is the main feature of NailGun.

**Comment by Ichimonji10** [comment]
> Other than the field definition would be required to have other information like workaround. For example many entities need to be wrapped in another hash, if that change from version to version?

These work-arounds could stay in `entities.py`. I'm proposing that only the field definitions move to schema files.

> If we use cumulative approach like migrations, then everytime nailgun start that should be computed.

Using a cumulative approach for defining fields would reduce redundancy. But a cumulative approach is also more fragile and complex. I'm not sure a cumulative approach is the right way to go.

> NailGun is smart enough to provide smart tools to work with the API by normalizing all differences and gotchas into a stable and clear API. We need to be careful here to not lose that as this is the main feature of NailGun.

Agreed. I also want to keep things fairly sane internally, too. Let's take our time and pick a "right" solution. NailGun isn't perfect as is, but it's serviceable enough that I'm in no rush to rewrite major portions of it.

**Comment by Ichimonji10** [comment]
Whenver this gets implemented, #220 should be revisited.

**Comment by renzon** [comment]
Right now both robottelo and nailgun are following "branch" approach. Thus fields do not need be dynamic. They could also been reverted back to Descriptor, making sintax cleaner and improving performance, once property would be created for each class instead of each instance.

Because of reasons above I am closing this issue in favor of #392
